ffcv/transforms/color_jitter.py

```diff
@@ -0,0 +1,139 @@
+
+            return images
+
+        brightness.is_parallel = True
+        return brightness
+
+    def declare_state_and_memory(self, previous_state):
+        return (replace(previous_state, jit_mode=True), AllocationQuery(previous_state.shape, previous_state.dtype))
+
+
+
+class RandomContrast(Operation):
+    '''
+    Randomly adjust image contrast. Operates on raw arrays (not tensors).
+
+    Parameters
+    ----------
+    magnitude : float
+        randomly choose contrast enhancement factor on [max(0, 1-magnitude), 1+magnitude]
+    p : float
+        probability to apply contrast
+    '''
+    def __init__(self, magnitude, p=0.5):
+        super().__init__()
+        self.p = p
+        self.magnitude = magnitude
+
+    def generate_code(self):
+        my_range = Compiler.get_iterator()
+        p = self.p
+        magnitude = self.magnitude
+
+        def contrast(images, *_):
+            def blend(img1, img2, ratio): return (ratio*img1 + (1-ratio)*img2).clip(0, 255).astype(img1.dtype)
+
+            apply_contrast = np.random.rand(images.shape[0]) < p
+            magnitudes = np.random.uniform(max(0, 1-magnitude), 1+magnitude, images.shape[0])
+            for i in my_range(images.shape[0]):
+                if apply_contrast[i]:
+                    r, g, b = images[i,:,:,0], images[i,:,:,1], images[i,:,:,2]
+                    l_img = (0.2989 * r + 0.587 * g + 0.114 * b).astype(images[i].dtype)
+                    images[i] = blend(images[i], l_img.mean(), magnitudes[i])
+
+            return images
+
+        contrast.is_parallel = True
+        return contrast
+
+    def declare_state_and_memory(self, previous_state):
+        return (replace(previous_state, jit_mode=True), AllocationQuery(previous_state.shape, previous_state.dtype))
+
+
+
+class RandomSaturation(Operation):
+    '''
+    Randomly adjust image color balance. Operates on raw arrays (not tensors).
+
+    Parameters
+    ----------
+    magnitude : float
+        randomly choose color balance enhancement factor on [max(0, 1-magnitude), 1+magnitude]
+    p : float
+        probability to apply saturation
+    '''
+    def __init__(self, magnitude, p=0.5):
+        super().__init__()
+        self.p = p
+        self.magnitude = magnitude
+
+    def generate_code(self):
+        my_range = Compiler.get_iterator()
+        p = self.p
+        magnitude = self.magnitude
+
+        def saturation(images, *_):
+            def blend(img1, img2, ratio): return (ratio*img1 + (1-ratio)*img2).clip(0, 255).astype(img1.dtype)
+
+            apply_saturation = np.random.rand(images.shape[0]) < p
+            magnitudes = np.random.uniform(max(0, 1-magnitude), 1+magnitude, images.shape[0])
+            for i in my_range(images.shape[0]):
+                if apply_saturation[i]:
+                    r, g, b = images[i,:,:,0], images[i,:,:,1], images[i,:,:,2]
+                    l_img = (0.2989 * r + 0.587 * g + 0.114 * b).astype(images[i].dtype)
```

Comment:
Where do these magic numbers come from ?

Reply:
Are they the same in pytorch ?

Reply:
As mentioned in reference, my codes are referenced based on torchvision transform and follows most of the main parts.

Magic number comes from [torchvision rgb to grayscale](https://github.com/pytorch/vision/blob/main/torchvision/transforms/functional_tensor.py#L140-L157), which is needed for [adjusting contrast](https://github.com/pytorch/vision/blob/main/torchvision/transforms/functional_tensor.py#L171-L185) and [adjusting saturation](https://github.com/pytorch/vision/blob/main/torchvision/transforms/functional_tensor.py#L217-L228).

PR ([#154](https://github.com/libffcv/ffcv/pull/154)) use similar number [(0.299, 0.587, 0.114)](https://github.com/libffcv/ffcv/pull/154/files#diff-6f067e767c6244a55051e3c87e3dc421dec4018ddaf63534bb0a45b4074ab193R109-R115) but it's not the same number as official torchvision number which is (0.2989, 0.587, 0.114).